lwt/happy_eyeballs_lwt.ml

```diff
@@ -46,21 +51,30 @@ let try_connect ip port =
 
 let rec act t action =
   let open Lwt.Infix in
+  let ( >>? ) = Lwt_result.bind in
   Log.debug (fun m -> m "[%u] action %a" t.counter
                 Happy_eyeballs.pp_action action);
   begin
     match action with
-    | Happy_eyeballs.Resolve_a host ->
-      begin
-        Dns_client_lwt.getaddrinfo t.dns Dns.Rr_map.A host >|= function
-        | Ok (_, res) -> Ok (Happy_eyeballs.Resolved_a (host, res))
-        | Error `Msg msg -> Ok (Happy_eyeballs.Resolved_a_failed (host, msg))
-      end
-    | Happy_eyeballs.Resolve_aaaa host ->
+    | Happy_eyeballs.Resolve_a host | Happy_eyeballs.Resolve_aaaa host ->
       begin
-        Dns_client_lwt.getaddrinfo t.dns Dns.Rr_map.Aaaa host >|= function
-        | Ok (_, res) -> Ok (Happy_eyeballs.Resolved_aaaa (host, res))
-        | Error `Msg msg -> Ok (Happy_eyeballs.Resolved_aaaa_failed (host, msg))
+        let record = match action with
+          | Happy_eyeballs.Resolve_a _ -> `A
+          | Happy_eyeballs.Resolve_aaaa _ -> `AAAA
+          | _ -> assert false (* never occur! *) in
+        Lwt.return (Domain_name.host host) >>?
+        t.getaddrinfo record >|= fun res -> match res, record with
+        | Ok set, `A ->
+          let fold ip set = match ip with
+            | Ipaddr.V4 ipv4 -> Ipaddr.V4.Set.add ipv4 set
+            | _ -> set in
```

Comment:
Should we log a warning? It seems relatively harmless, but it could be an annoying bug to track down if you mix up A and AAAA in your getaddrinfo implementation.